zict/async_buffer.py

```diff
@@ -0,0 +1,157 @@
+
+    def close(self) -> None:
+        # Call LRU.close(), which stops LRU.evict_until_below_target() halfway through
+        super().close()
+        if self.executor is not None:
+            # TODO just use shutdown(cancel_futures=True) (requires Python >=3.9)
+            for future in self.futures:
+                future.cancel()
+            self.executor.shutdown(wait=True)
+            self.executor = None
+
+    def _offload(self, func: Callable[..., T], *args: Any) -> asyncio.Future[T]:
+        if self.executor is None:
+            self.executor = ThreadPoolExecutor(
+                1, thread_name_prefix="zict.Buffer offloader"
+            )
+
+        loop = asyncio.get_running_loop()
+        context = contextvars.copy_context()
+        future = loop.run_in_executor(self.executor, context.run, func, *args)
+        self.futures.add(future)
+        future.add_done_callback(self.futures.remove)
+        return future  # type: ignore[return-value]
+
+    # Return an asyncio.Future, instead of just writing it as an async function, to make
+    # it easier for overriders to tell apart the use case when all keys were already
+    # in fast
+    @locked
+    def async_get(
+        self, keys: Collection[KT], missing: Literal["raise", "omit"] = "raise"
+    ) -> asyncio.Future[dict[KT, VT]]:
+        """Fetch one or more key/value pairs. If not all keys are available in fast,
+        offload to a worker thread moving keys from slow to fast, as well as possibly
+        moving older keys from fast to slow.
+
+        Parameters
+        ----------
+        keys:
+            collection of zero or more keys to get
+        missing: raise or omit, optional
+            raise (default)
+                If any key is missing, raise KeyError.
+            omit
+                If a key is missing, return a dict with less keys than those requested.
+
+        Notes
+        -----
+        All keys may be present when you call ``async_get``, but ``__delitem__`` may be
+        called on one of them before the actual data is fetched. ``__setitem__`` also
+        internally calls ``__delitem__`` in a non-atomic way, so you may get
+        ``KeyError`` when updating a value too.
+        """
+        # This block avoids spawning a thread if keys are missing from both fast and
+        # slow. It is otherwise just a performance optimization.
+        if missing == "omit":
+            keys = [key for key in keys if key in self]
+        elif missing == "raise":
+            for key in keys:
+                if key not in self:
+                    raise KeyError(key)
+        else:
+            raise ValueError(f"missing: expected raise or omit; got {missing}")
+        # End performance optimization
+
+        try:
+            # Do not pull keys towards the top of the LRU unless they are all available.
+            # This matters when there is a very long queue of async_get futures.
+            d = self.fast.get_all_or_nothing(keys)
```

Comment:
Strictly speaking, fast doesn't need to be a LRU, does it? In this case, I suspect we'll get an `AttributeError` here, won't we?

Reply:
Buffer.fast is always an LRU:
https://github.com/dask/zict/blob/b96afc4158c8b129e1729b75def67d87982dc41c/zict/buffer.py#L79-L85